Domain/Sia.Core/Protocol/PaginatedLinksHeader.cs

```diff
@@ -0,0 +1,66 @@
+﻿namespace Sia.Core.Protocol
+{
+    using Sia.Core.Data;
+    using Sia.Core.Validation;
+    using System.Collections.Generic;
+    using System.Linq;
+
+    public class PaginatedLinksHeader : ILinksHeader
+    {
+        private IFilterMetadataProvider _filterMetadata { get; }
+        private PaginationMetadata _pagination { get; }
+        private string _baseRoute { get; }
+        private OperationLinks _operationLinks { get; }
+        private RelationLinks _relationLinks { get; }
+
+        public PaginatedLinksHeader(
+            string baseRoute,
+            OperationLinks operationLinks,
+            RelationLinks relationLinks,
+            PaginationMetadata pagination,
+            IFilterMetadataProvider filterMetadata = null
+        )
+        {
+            _operationLinks = ThrowIf.Null(operationLinks, nameof(operationLinks));
+            _relationLinks = relationLinks;
+            _baseRoute = ThrowIf.NullOrWhiteSpace(baseRoute, nameof(baseRoute));
+            _pagination = ThrowIf.Null(pagination, nameof(pagination));
+            _filterMetadata = filterMetadata ?? new NoFilterMetadata();
+        }
+
+        public const string HeaderName = "links";
+
+        public LinksForSerialization GetHeaderValues()
+        {
+            var toReturn = new LinksForSerialization
+            {
+                Metadata = new LinksMetadata()
+                {
+                    Pagination = _pagination.ToSerializableValues()
+                },
+                Links = new LinksCollection()
+                {
+                    Operations = _operationLinks,
+                    Pagination = new PaginationLinks()
+                    {
+                        Previous = MakeDirectionalLink(_pagination.Previous),
+                        Next = MakeDirectionalLink(_pagination.Next)
+                    },
+                    Related = _relationLinks
+                }
+            };
+            return toReturn;
+        }
+
+        protected string MakeDirectionalLink(DirectionalPaginationMetadata page)
+            => page.Exists
+                ? _baseRoute + FormatUrl(page.LinkInfo.Concat(_filterMetadata.FilterValues()))
+                : null;
+
+        protected static string UrlTokenFormat(KeyValuePair<string, string> token)
+            => $"{token.Key}={token.Value}";
+
+        protected static string FormatUrl(IEnumerable<KeyValuePair<string, string>> tokens)
+            => "/?" + string.Join("&", tokens.Select(UrlTokenFormat));
```

Comment:
Keys and values in url strings should be url encoded... One should normally suggest a helper class like [`QueryHelper`](https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNetCore.WebUtilities/QueryHelpers.cs#L63) but in this case it forces your `IEnumerable<KeyValuePair<string, string>>` to an `IDictionary<string, string>` (dictionaries must have all unique keys, so you'll lose array support and have to do extra work to ensure unique keys if you go that route)

To that end it might be best to lift the good practices from `QueryHelper.AddQueryString`: 
```
        protected static string UrlTokenFormat(KeyValuePair<string, string> token)
            => $"{UrlEncoder.Default.Encode(token.Key)}={UrlEncoder.Default.Encode(token.Value)}";
```


...To be fair to `QueryHelper` you could call `AddQueryString(string, string, string)` repeatedly... but I'm far less happy with how that looks at first cut: 
```
        protected string MakeDirectionalLink(DirectionalPaginationMetadata page)
        {
            if (!page.Exists) 
            {
               return null;
            }

            string directionalLink = _baseRoute; 
            foreach (var filterValue in _filterMetadata.FilterValues())
            {
                directionalLink = QueryHelper.AddQueryString(directionalLink, filterValue.Key, filterValue.Value);
            }
            return directionalLink;
        }
```